### ISSUE TITLE
Remove unused CSS needs further consideration

### ISSUE THREAD

**Issue by peixotorms** [GoogleChrome/web.dev#1102]
I've been scratching my head around the opportunity named "Remove unused CSS" under the pagespeed insights web tool, and I think there should be a discussion about this metric:

**The opportunity description**
Currently, the "Remove unused CSS" states that _Remove dead rules from stylesheets and defer the loading of CSS not used for above-the-fold content to reduce unnecessary bytes consumed by network activity._
It then points to [](https://developers.google.com/web/tools/lighthouse/audits/unused-css) which basically only talks about deferring unused CSS.

These are two different things and the description (and possibly the metric itself) needs changes.
If the metric is for "Remove unused CSS" then there should be no information at all, regarding deferring CSS. 
The document for the unused-css page, basically tells us to extract the critical path, inline it on the header and defer the remaining CSS code... however, doing that will not pass the "Remove unused CSS" opportunity.
The only thing that will pass the "Remove unused CSS" opportunity, is to "remove dead rules from stylesheets" (nothing to do with deferring or the critical path), so developers will spend hours trying to defer CSS and extract the critical path, while still seeing this opportunity.
All information about deferring CSS should be deleted from the "Remove unused CSS" and moved into the "Eliminate render-blocking resources", since there is where it really fits.

Leaving only an explanation to "Remove dead rules from stylesheets" under "Remove unused CSS" is self-explanatory and doesn't require a documentation page for that... however, enter the 2KB limit.

**Some background info:**
When working with wordpress sites, very often we have plenty of css files. 
We can automate the merging of CSS files using a plugin, so the merged CSS file will include both used and unused css rules on that page.

**The 2Kb limit**
The "Remove unused CSS" metric stands correct in these cases, because the resulting CSS file usually has a lot of rules which are unused on that page. However, any CSS file with more than 2KB on unused CSS code (on that page), will trigger this opportunity and affect the score significantly, in some cases.

I think that the hard limit of 2 KB to trigger this opportunity should change to a relative size, rather than an absolute value. In addition, a total network payload size for the CSS should be implemented.

How I would trigger this metric, would be something like this:

a) if the total CSS is less than 125 Kb compressed (more on this later), do nothing (pass).
b) if the total CSS is more than 125 Kb, consider a relative ratio of unused/total size.

If the CSS file is more than 125 Kb, the relative size of the unused rules should affect scores in ranges proportionally... ex: if the CSS has more than 80% of unused CSS, it should shave away more points on the metric score than if it has less than 50% and again, less than 20% unused CSS.

The reason why I defend that the metric should do nothing if the CSS is less than 125 Kb, is because sometimes more is less.

**The browser cache**

If we grab all possible CSS rules across the website (used and unused on that page), we will end up with a considerable big CSS file, which for the sake of this metric, will have dead code in it.
For that page alone, the code is unused... however, that file includes rules used across the website and therefore, it makes sense to be there.

Generating a larger CSS file once and enqueueing it (async/defer), will store it on the browser cache and it can be reused across the site for multiple pageviews.
It's true that the initial network payload will be higher, but it get's diluted with more pageviews.

If developers have to remove unused CSS code from their CSS files (with only a 2KB margin), what this will cause is that every page will need it's own CSS file, which will have to be downloaded per pageview, even if some of that code is repeated.

What this causes ultimately, is that while each page may have a smaller individual network load, the amount of data downloaded if the user visits multiple pages will increase. 
If the goal was to save data for the mobile user, we need to consider that while this metric is applicable for one page designs, it doesn't really apply for large sites and for repeating pageviews (unless specific scenarios). 

Moreover, even if the developer removed all unused CSS code from the site, the chances of the final CSS file having only less than 2KB of unused CSS on all pages, is nearly zero.
In order to take full advantage of browser cache, some unused css allowance (within reasonable) should be given, to pass this metric opportunity. 

The only issue here, would be the 2KB limit being too strict without considering total CSS size or the ratio of used/unused code if it goes over the limit. 

It should not matter much if 80% of 50 Kb is unused (if we know it's used somewhere else), or perhaps if 10 to 20% out of 150 Kb is unused.

In my opinion, this needs further discussion by the team because I've seen developers creating one CSS file per page, just to pass this metric, thus eliminating any browser cache performance.

Clients care more about the metric number, than the actual speed... so if the goal is speed, there is an obligation from google to discuss this rather than simply focusing on one page tests.

So in short, 

a) The "Remove unused CSS" should have no reference to the critical path + defer CSS (because it doesn't allow you to pass that metric even if you do that).
b) The critical path + defer info, should be moved to the "Eliminate render-blocking resources", since there is more appropriate
c) The 2 KB limit to trigger the "Remove unused CSS" needs further discussion.

Thanks

**Comment by robdodson** [comment]
I tweaked your initial issue a bit to move the summary up to the top. Hope that's ok.

Moving this over to the Lighthouse repo for further discussion.
